File: app/src/main/java/social/entourage/android/api/model/TimestampedObject.java
package social.entourage.android.api.model;

import com.google.gson.annotations.Expose;

import java.util.Comparator;
import java.util.Date;

/**
 * Created by mihaiionescu on 26/02/16.
 */
public abstract class TimestampedObject {

    public static final int SEPARATOR = 0;
    public static final int CHAT_MESSAGE_ME = 1;
    public static final int CHAT_MESSAGE_OTHER = 2;
    public static final int TOUR_USER_JOIN = 3;
    public static final int TOUR_STATUS = 4;
    public static final int ENCOUNTER = 5;
    public static final int TOUR_CARD = 6;
    public static final int ENTOURAGE_CARD = 7;
    public static final int FEED_MEMBER_CARD = 8;
    public static final int INVITATION_CARD = 9;
    public static final int DATE_SEPARATOR = 10;
    public static final int TOP_VIEW = 998;
    public static final int BOTTOM_VIEW = 999;

    @Expose(serialize = false)
    private int hashCode;

    public abstract Date getTimestamp();

    public abstract String hashString();

    @Override
    public int hashCode() {
        if (hashCode == 0) {
            hashCode = hashString().hashCode();
        }
        return hashCode;
    }

    public abstract int getType();

    public abstract long getId();

    /**
     * Copies the local fields (the ones not retrieved from the server) from other object
     * @param other The object to copy from
     */
    public void copyLocalFields(TimestampedObject other) {}


    // ----------------------------------
    // INNER CLASSES
    // ----------------------------------

    public static class TimestampedObjectComparatorOldToNew implements Comparator<TimestampedObject> {
        @Override
        public int compare(final TimestampedObject lhs, final TimestampedObject rhs) {
            if (lhs.getTimestamp() != null && rhs.getTimestamp() != null) {
                Date date1 = lhs.getTimestamp();
                Date date2 = rhs.getTimestamp();
                return date1.compareTo(date2);
            } else {
                return 0;
            }
        }
    }
}

File: app/src/main/java/social/entourage/android/api/model/TimestampedObject.kt
package social.entourage.android.api.model

import com.google.gson.annotations.Expose
import java.util.*

/**
 * Created by mihaiionescu on 26/02/16.
 */
abstract class TimestampedObject {
    @Expose(serialize = false)
    private var hashCode = 0
    abstract val timestamp: Date?
    abstract fun hashString(): String
    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode = hashString().hashCode()
        }
        return hashCode
    }

    abstract val type: Int
    abstract val id: Long

    /**
     * Copies the local fields (the ones not retrieved from the server) from other object
     * @param other The object to copy from
     */
    open fun copyLocalFields(other: TimestampedObject) {}

    // ----------------------------------
    // INNER CLASSES
    // ----------------------------------
    class TimestampedObjectComparatorOldToNew : Comparator<TimestampedObject> {
        override fun compare(lhs: TimestampedObject, rhs: TimestampedObject): Int {
            return if (rhs.timestamp != null) {
                lhs.timestamp?.compareTo(rhs.timestamp) ?: 0
            } else {
                0
            }
        }
    }

    companion object {
        const val SEPARATOR = 0
        const val CHAT_MESSAGE_ME = 1
        const val CHAT_MESSAGE_OTHER = 2
        const val USER_JOIN = 3
        const val ENTOURAGE_CARD = 7
        const val FEED_MEMBER_CARD = 8
        const val INVITATION_CARD = 9
        const val DATE_SEPARATOR = 10
        const val GUIDE_POI = 11
        const val INVITATION_LIST = 12
        const val ANNOUNCEMENT_CARD = 13
        const val CHAT_MESSAGE_OUTING = 14
        const val STATUS_UPDATE_CARD = 15
        const val LOADER_CARD = 16
        const val TOP_VIEW = 998
        const val BOTTOM_VIEW = 999
    }
}